components/board/client.go

```diff
@@ -253,6 +265,125 @@ func (dic *digitalInterruptClient) RemoveCallback(c chan Tick) {
 	panic(errUnimplemented)
 }
 
+func (c *client) StreamTicks(ctx context.Context, interrupts []string, ch chan Tick, extra map[string]interface{}) error {
+	c.streamMu.Lock()
+	defer c.streamMu.Unlock()
+	ext, err := protoutils.StructToStructPb(extra)
+	if err != nil {
+		return err
+	}
+	c.extra = ext
+
+	// Start the stream
+	c.streamRunning = true
+	c.activeBackgroundWorkers.Add(1)
+	ctx, cancel := context.WithCancel(ctx)
+	c.cancelBackgroundWorkers = cancel
+	// Create a background go routine that connects to the server's stream
+	utils.PanicCapturingGo(func() {
+		defer c.activeBackgroundWorkers.Done()
+		c.connectTickStream(ctx, interrupts, ch)
+	})
+	c.mu.RLock()
+	ready := c.streamReady
+	c.mu.RUnlock()
+
+	// wait until the stream is ready to return
+	for !ready {
+		c.mu.RLock()
+		ready = c.streamReady
+		c.mu.RUnlock()
+		// wait for 50 ms before checking again
+		if !utils.SelectContextOrWait(ctx, 50*time.Millisecond) {
+			return ctx.Err()
```

Comment:
If the context is cancelled, I suspect we should also shut down our background worker, though I could be wrong about that.